### ISSUE TITLE
Numba appears ~10X slower running in docker versus native on OSX

### ISSUE THREAD

**Issue by michaelbarton** [numba/numba#8293]
TLDR: Running the same numba code in docker versus on OSX appears almost ~10X slower using timeit benchmarks.

- Numba on OSX: 0.678s runtime
- Numba in docker: 4.900s runtime

(N=10000)

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Hi! I really enjoy numba, and appreciate all the work that goes into it. I wanted to report a strange phenomenon that I thought might be noteworthy. I'm doing some numerical simulation using multinomial distributions. I'm doing millions of samplings and so calculation of a `logpmf` is becoming a bottleneck. I wanted to give numba a try to see if there was any speed up possible. I've attempted to rewrite the hot code paths in numba to see if there were any potential gains there. I however notice the same code running on OSX runs much slower when running in a docker image. The code needs to run in a docker container eventually because it will be deployed in a k8s cluster. Perhaps I'm not writing the code in the most optimal way, but it does seem strange that running in docker could cause such a dramatic change in performance. I might expect a few percentage points because of virtualisation, but not this much. I've attached a reproducible example below.

Potential docker considerations:
- Using an intel mac, not arm.
- Docker VM has 10CPUs available vs 12 for the OSX.
- Docker VM has 1GB of swap available.

```python
import ctypes
import timeit

import numba
import numpy
from numba import extending
from numpy.typing import NDArray

_PTR = ctypes.POINTER
_dble = ctypes.c_double
_ptr_dble = _PTR(_dble)

gammaln_functype = ctypes.CFUNCTYPE(_dble, _dble)
cython_gammaln = gammaln_functype(
    extending.get_cython_function_address("scipy.special.cython_special", "gammaln")
)

xlogy_functype = ctypes.CFUNCTYPE(_dble, _dble, _dble)
cython_xlogy = xlogy_functype(
    extending.get_cython_function_address("scipy.special.cython_special", "__pyx_fuse_1xlogy")
)


@numba.vectorize([numba.float64(numba.float64)])
@numba.jit(nopython=True, parallel=True, fastmath=True)
def numba_gammaln(x):
    return cython_gammaln(x)


@numba.vectorize([numba.float64(numba.float64, numba.float64)])
@numba.jit(nopython=True, parallel=True, fastmath=True)
def numba_xlogy(x, y):
    return cython_xlogy(x, y)


@numba.jit(nopython=True, parallel=True, fastmath=True)
def numba_logpmf(x: NDArray[numpy.int_], n: int, p: NDArray[numpy.float32]) -> float:
    """Calculate the log probability mass function using vectorised scipy special functions."""
    difference = numpy.sum(numba_xlogy(x, p) - numba_gammaln(x + 1))
    return cython_gammaln(n + 1) + difference


#
# Benchmark
#

N_TESTS = 10000

n = 10
p = numpy.array([0.08333333, 0.08333333, 0.08333333, 0.08333333, 0.83333333])
x = numpy.array([1, 1, 1, 1, 6])

# Trigger JIT if required
numba_logpmf(x, n, p)

print(timeit.timeit(lambda: numba_logpmf(x, n, p), number=N_TESTS))
```

**Comment by esc** [comment]
@michaelbarton thank you for submitting this.

> * Using an intel mac, not arm.

Using an intel mac with an x86_64 based docker container I get:

Native: 0.07741974503733218
Docker: 0.04650230007246137

So it's actually faster in docker for me. What kind of mac are you using? What kind of docker container are you using?

**Comment by michaelbarton** [comment]
Hi esc, here are the details of my system:

```
  Model Name:	MacBook Pro
  Model Identifier:	MacBookPro16,1
  Processor Name:	6-Core Intel Core i7
  Processor Speed:	2.6 GHz
  Number of Processors:	1
  Total Number of Cores:	6
  L2 Cache (per Core):	256 KB
  L3 Cache:	12 MB
  Hyper-Threading Technology:	Enabled
  Memory:	32 GB
  System Firmware Version:	1731.140.2.0.0 (iBridge: 19.16.16064.0.0,0)
```

Here's the output of `docker info`

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.7.1)
  compose: Docker Compose (Docker Inc., v2.2.1)
  scan: Docker Scan (Docker Inc., 0.9.0)

Server:
 Containers: 60
  Running: 50
  Paused: 0
  Stopped: 10
 Images: 50
 Server Version: 20.10.11
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runtime.v1.linux runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc version: v1.0.2-0-g52b36a2
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.10.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 10
 Total Memory: 22.5GiB
 Name: docker-desktop
 ID: 7UWX:4T4O:QESQ:GAEN:BQJR:DNH2:XBWE:MYMM:KQBY:BE7W:FPEN:Q5PS
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

```

**Comment by michaelbarton** [comment]
I'm using the `python:3.10-slim` docker image to run the tests, I'm installing the software with a poetry venv.

```
root@f9628f4c717a:~# env | sort
GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D
HOME=/root
HOSTNAME=f9628f4c717a
LANG=C.UTF-8
PATH=/opt/poetry/bin:/opt/pysetup/.venv/bin:/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PIP_NO_CACHE_DIR=1
PLAT=linux-x86_64
POETRY_CACHE_DIR=/opt/poetry/.cache
POETRY_HOME=/opt/poetry
POETRY_NO_INTERACTION=1
PWD=/root
PYSETUP_PATH=/opt/pysetup
PYTEST_TMP_DIR=/mnt/tmp
PYTHONUNBUFFERED=1
PYTHON_GET_PIP_SHA256=ba3ab8267d91fd41c58dbce08f76db99f747f716d85ce1865813842bb035524d
PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/6ce3639da143c5d79b44f94b04080abf2531fd6e/public/get-pip.py
PYTHON_PIP_VERSION=22.0.4
PYTHON_SETUPTOOLS_VERSION=58.1.0
PYTHON_VERSION=3.10.5
SHLVL=0
TERM=xterm
```

**Comment by michaelbarton** [comment]
I rebooted docker desktop and reran the test, I got a higher number this time: `6.838`
So perhaps something specific to my system causing this issue?

**Comment by michaelbarton** [comment]
Ok, I found the issue, I switched the base image from `python:3.10-slim` to `python:3.10`, the performance was `0.069724`.
I assume there might be an apt package that's not present in `python:3.10-slim` that could be required? Could you guess at what that might be?

**Comment by esc** [comment]
> Ok, I found the issue, I switched the base image from `python:3.10-slim` to `python:3.10`, the performance was `0.069724`. I assume there might be an apt package that's not present in `python:3.10-slim` that could be required? Could you guess at what that might be?

I am glad you managed to resolve this issue. I have no idea what package could be missing that would incur such a performance penalty.

**Comment by rez0n** [comment]
@michaelbarton Thanks for this issue, same thing.

**Comment by esc** [comment]
> @michaelbarton Thanks for this issue, have similar problem. Update: Using `python:3.10` does not solved the problem.

is it worth opening a new issue about this?

**Comment by rez0n** [comment]
Hi @esc 
I'm pretty sure yes, as most probably it wide range issue.
But I am not sure that I am able to provide detailed information about this, I use `timezonefinder` (which uses numba under the hood) and it become process ~400 entities from 1 second to 40 in Docker after upgrade from 0.54.0 to 0.56.0.

**Comment by esc** [comment]
@rez0n thank you for checking.

**Comment by rez0n** [comment]
Hi @esc 
The Timezonefinder issue had updates.
In short debugging shows that the module use Numba but it doesn't give any performance profit, so it means something wrong with Numba (or environment), I have no more ideas on how to solve this.

**Comment by esc** [comment]
> Hi @esc The Timezonefinder issue had updates. In short debugging shows that the module use Numba but it doesn't give any performance profit, so it means something wrong with Numba (or environment), I have no more ideas on how to solve this

Hmmm, perhaps it makes sense to check if Numba is beneficial to Timezonefinder outside of docker?

**Comment by jannikmi** [comment]
> Hmmm, perhaps it makes sense to check if Numba is beneficial to Timezonefinder outside of docker?

It is. 

When i am running the tests without Numba:
`98 passed in 105.28s (0:01:45)`

with Numba `numba-0.55.2-cp38-cp38-macosx_10_14_x86_64.whl` installed:
`98 passed in 19.04s`

**Comment by jannikmi** [comment]
I have the gut feeling that the difference is measured speed has something to do with the initial time required by Numba to just in time compile the functions (which afterwards are being cached).

I guess due to the docker setup the cached JIT compiled functions will be forgotten each time, while in your local setup they will be reused (giving you the speed benefit).

**Comment by rez0n** [comment]
Meanwhile I not see any reason to "forget compiled functions" in Docker. 
it is stored in __pycache__? 
I running 4 testing attempts on same dataset (just hitting F5 for api endpoint which use timezonefinder under the hood) so any cache (if it has been created should be in place)
